src/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs

```diff
@@ -0,0 +1,103 @@
+
+        public NativeLibrary(string libraryName, IntPtr handle)
+        {
+            Name = libraryName;
+            Handle = handle;
+        }
+
+        static NativeLibrary()
+        {
+            AssemblyToCallbackMap = new ConditionalWeakTable<Assembly, Func<LoadNativeLibraryArgs, NativeLibrary>>();
+        }
+
+        // It is called by a user per assembly: RegisterNativeLibraryLoadCallback(assembly, customMappingCallback)
+        // triggered by an event on AppDomain
+        public static bool RegisterNativeLibraryLoadCallback(Assembly assembly, Func<LoadNativeLibraryArgs, NativeLibrary> callback)
+        {
+            Func<LoadNativeLibraryArgs, NativeLibrary> registeredCallback = null;
+            bool callbackAlreadyRegistered = AssemblyToCallbackMap.TryGetValue(assembly, out registeredCallback);
+
+            if (!callbackAlreadyRegistered)
+            {
+                AssemblyToCallbackMap.Add(assembly, callback);
+            }
+            else
+            {
+                AssemblyToCallbackMap.Remove(assembly);
+                AssemblyToCallbackMap.Add(assembly, callback);
+            }
+
+            return true;
+        }
+
+        // A wrapper function for Load(string libraryName, DllImportSearchPath dllImportSearchPath, Assembly assembly)
+        public static NativeLibrary Load(string libraryName)
+        {
+            return Load(libraryName, DllImportSearchPath.LegacyBehavior, null);
+        }
+
+        public static NativeLibrary Load(string libraryName, DllImportSearchPath dllImportSearchPath, Assembly assembly)
+        {
+            RuntimeAssembly assemblyAsRuntimeAssembly = null;
+            if (assembly != null)
+            {
+                assemblyAsRuntimeAssembly = assembly as RuntimeAssembly;
+            }
+
+            // Todo: Determine when it shouldn't be true
+            bool searchAssemblyDirectory = true;
+
+            IntPtr hmodule = LoadLibrary(assemblyAsRuntimeAssembly, libraryName, searchAssemblyDirectory, (int)dllImportSearchPath);
+
+            NativeLibrary loadedLibrary = new NativeLibrary(libraryName, hmodule);
+
+            return loadedLibrary;
+        }
+
+        // Calls NativeLibrary::LoadLibrary unmanaged methode
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        internal static extern IntPtr LoadLibrary(RuntimeAssembly assembly, string libraryName, bool searchAssemblyDirectory, int dllImportSearchPathFlag);
+
+        // Todo: optimize for methods that do not have callbacks
+        internal static IntPtr LoadLibraryCallback(string libraryName, uint dllImportSearchPathUint, Assembly assembly)
+        {
+            DllImportSearchPath dllImportSearchPath = (DllImportSearchPath)dllImportSearchPathUint;
+
+            Func<LoadNativeLibraryArgs, NativeLibrary> callback = null;
+            AssemblyToCallbackMap.TryGetValue(assembly, out callback);
+
+            if (callback != null)
```

Comment:
Bah.... The semantics of `TryGetValue()` should only return `true` when the callback is non-null (i.e. has a value).

`if (AssemblyToCallbackMap.TryGetValue(assembly, out callback)) { /* callback not null/was found */ }`